### ISSUE TITLE
Cracking doesnt work correctly with multiple CPUs

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#109]
**Reported by ASPj on 16 Feb 2007 14:25 UTC**

I collect approximately 1.1 million IVs and run aircrack with all default options and after about 10 seconds I get "Out of luck key not found etc", except the key values displayed are correct! If I force single processor cracking (-p 1) for aircrack-ng then the key is found with no problem.

**Comment by aircrack-ng** [comment]
**Comment by gandalf on 16 Feb 2007 14:25 UTC**

Replying to [ASPj](ticket:109):
> I collect approximately 1.1 million IV's and run aircrack with all defualt options and after about 10 seconds I get Out of luck key not found etc, except the key values displayed are correct! If i force single processor cracking (-p 1) , for aircrack-ng then they key is found with no problem.

Could you please put the dump online, so I can check on my Pentium D if the same problem occurs ?

Thanks.

**Comment by aircrack-ng** [comment]
**Comment by ThiefMaster on 16 Feb 2007 14:25 UTC**

I can confirm that: http''':'''//tm.privatepaste.com/de01rEw87H

PS: This URL blocker is annoying.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2007 14:25 UTC**

Taken from http://tm.privatepaste.com/de01rEw87H:

1 CPU:

```
# aircrack-ng *.ivs -c -s -b 00:16:B6:C2:BA:76 -p 1 -d '61:61:61:61:61:61:61:61:61:61:61:61:61'
Opening test-03.ivs
Opening test-04.ivs
Read 1244797 packets.

                                                                        Aircrack-ng 0.7

                                                          [Tested 99 keys (got 1244796 IVs)
  KB    depth   byte(vote)
    0    0/  1   61(+inf) 6A(  30) 44(  10) 2C(   3) 64(   3) 2E(   0) 37(   0) 42(   0) 46(   0) 47(   0) 4A(   0) 59(   0) 60(   0) 01(  -1) 02(  -1)   a
    1    0/  1   61(+inf) 52(  28) 79(  17) 77(  13) 62(   7) 69(   5) 2E(   0) 40(   0) 41(   0) 49(   0) 01(  -1) 02(  -1) 03(  -1) 04(  -1) 05(  -1)   a
    2    0/  1   61(+inf) 21(  18) 31(  12) 3B(  12) 77(  12) 79(  12) 7C(   6) 27(   0) 28(   0) 2B(   0) 2C(   0) 2F(   0) 30(   0) 47(   0) 4C(   0)   a
    3    0/  1   61(+inf) 38(  30) 73(  15) 74(  15) 27(   0) 2B(   0) 2C(   0) 2D(   0) 32(   0) 4E(   0) 4F(   0) 51(   0) 7C(   0) 01(  -1) 02(  -1)   a
    4    0/  1   61(+inf) 4C( 135) 62( 111) 34(  27) 52(  27) 29(  15) 60(  15) 78(  15) 42(   7) 28(   3) 5C(   3) 27(   0) 58(   0) 59(   0) 5D(   0)   a
    5    0/  1   61(+inf) 56(  15) 67(  15) 7E(  15) 43(   5) 58(   5) 3F(   3) 5A(   3) 23(   0) 24(   0) 25(   0) 2B(   0) 34(   0) 41(   0) 44(   0)   a
    6    0/  1   61(+inf) 79(  39) 55(  33) 4A(  15) 56(  15) 54(   5) 6F(   1) 20(   0) 21(   0) 38(   0) 39(   0) 41(   0) 52(   0) 53(   0) 01(  -1)   a
    7    0/  1   61(+inf) 4D( 180) 4C(  15) 4F(  15) 5F(  15) 42(  12) 50(  12) 6A(   5) 70(   4) 22(   3) 20(   0) 21(   0) 23(   0) 24(   0) 2E(   0)   a
    8    0/  1   61(+inf) 46(  15) 7E(  15) 3C(   5) 7A(   3) 00(   0) 20(   0) 26(   0) 3D(   0) 3F(   0) 47(   0) 48(   0) 49(   0) 4A(   0) 62(   0)   a
    9    0/  1   61(+inf) 3B(  72) 35(  33) 44(  27) 4B(  12) 77(  12) 41(   3) 71(   3) 75(   3) 40(   0) 42(   0) 43(   0) 46(   0) 47(   0) 48(   0)   a
   10    0/  1   61(+inf) 62(  24) 2B(   5) 38(   5) 39(   5) 4B(   5) 64(   5) 6E(   5) 27(   3) 7C(   3) 23(   0) 25(   0) 28(   0) 37(   0) 3A(   0)   a
   11    0/  1   61(+inf) 6E(  60) 58(  57) 67(  57) 2F(  40) 5E(  39) 38(  32) 6A(  30) 2E(  18) 63(  15) 35(   8) 4B(   7) 32(   5) 62(   3) 64(   3)   a

     KEY FOUND! [ 61:61:61:61:61:61:61:61:61:61:61:61:61 ](00:00:02]) (ASCII: aaaaaaaaaaaaa )
```

2 CPU:
```
# aircrack-ng *.ivs -c -s -b 00:16:B6:C2:BA:76 -d '61:61:61:61:61:61:61:61:61:61:61:61:61'
Opening test-03.ivs
Opening test-04.ivs
Read 1244797 packets.

   Attack failed. Possible reasons:
      * Out of luck: you must capture more IVs. Usually, 104-bit WEP
        can be cracked with about one million IVs, sometimes more.
      * If all votes seem equal, or if there are many negative votes,
        then the capture file is corrupted, or the key is not static.
      * A false positive prevented the key from being found.  Try to
        disable each korek attack (-k 1 .. 17), raise the fudge factor
        (-f)
```

**Comment by aircrack-ng** [comment]
**Modified by darkAudax on 16 Feb 2007 14:25 UTC**

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 16 Feb 2007 14:25 UTC**

the way aircrack-ng reads the dictionary and serves the cracking threads is flawed by design and may be rewritten as a whole.

**Comment by aircrack-ng** [comment]
**Modified by misterx on 16 Feb 2007 14:25 UTC**

**Comment by aircrack-ng** [comment]
**Comment by nx5 on 16 Feb 2007 14:25 UTC**

The last keybytes bruteforcing may be different for multiproc and single proc. As such, you may crack or not a key depending on that. In the particular case of this example, it looks like there could be a problem, though. Can somebody upload the .ivs files somewhere and provide a link?


Reference about the multiple processors cracking in aircrack: http.....//tinyshell.be/aircrackng/forum/index.php?topic=124.msg541;topicseen#msg541

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2007 14:25 UTC**

WPA cracking should be fixed in aircrack-ng/aircrack-ng@797d92a821d963770bfc4d59db43efeb734448cf

**Comment by aircrack-ng** [comment]
**Comment by mb on 16 Feb 2007 14:25 UTC**

Replying to [ASPj](ticket:109):
> I collect approximately 1.1 million IVs and run aircrack with all default options and after about 10 seconds I get "Out of luck key not found etc", except the key values displayed are correct! If I force single processor cracking (-p 1) for aircrack-ng then the key is found with no problem.

I see cracking being broken with latest aircrack-ng release on my 4way machine, too.
It doesn't utilize CPU correctly and it generates _high_ system load on each thread. In total it only uses about 140% of the available 400% CPU power.
In an earlier version (that one shipped by Ubuntu) it used to work.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 16 Feb 2007 14:25 UTC**

Replying to [mb@bu3sch.de](comment:9):
> I see cracking being broken with latest aircrack-ng release on my 4way machine, too.
> It doesn't utilize CPU correctly and it generates _high_ system load on each thread. In total it only uses about 140% of the available 400% CPU power.
> In an earlier version (that one shipped by Ubuntu) it used to work.

ah, forgot to say that with -p 1 it works correctly.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2007 14:25 UTC**

Which version was shipped with your ubuntu?

**Comment by aircrack-ng** [comment]
**Comment by System14 on 16 Feb 2007 14:25 UTC**

Just some additional information. Ubuntu 7.04 ships with aircrack-ng_0.6-1 which runs very fast on my duo core using near %200 cpu usage. I d/l 0.9.1 and get only %105 with the default run (-p2). Using -p1 aircrack-ng uses near 100% cpu time. Using -p1 i can crack the test file in half the time than using -p2

**Comment by aircrack-ng** [comment]
**Comment by hirte on 16 Feb 2007 14:25 UTC**

could someone retry the current 1.0 dev version (>= aircrack-ng/aircrack-ng@), korek ('-K') should be fixed on multiple cpus. if the load problem is still there, i'd like to know if it worked with 0.7, 0.8 and 0.9 and at which point the problem got introduced.

**Comment by aircrack-ng** [comment]
**Comment by hvidevold on 16 Feb 2007 14:25 UTC**


I can confirm this bug with WPA PSK cracking using aircrack-ng from SVN revision 1166 (Aug 26 2008). On my dual core system cracking works with -p 1 option but not without.

Works:
aircrack-ng -p 1 -a 2 -w testwords.txt -b 00:14:51:78:45:D9 airodump-06.cap

Doesn't work:
aircrack-ng -a 2 -w testwords.txt -b 00:14:51:78:45:D9 airodump-06.cap

Here is the cap file:
http ://rasmus.uib.no/~ehv022/aircrack-bug/airodump-06.cap

Here is the wordlist with the correct password on a single line:
http ://rasmus.uib.no/~ehv022/aircrack-bug/testwords.txt

I'll attach the files also.

**Comment by aircrack-ng** [comment]
**Attachment by hvidevold on 26 Aug 2008 02:34 UTC**

Uploaded file [airodump-06.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20080826-023450_000-airodump-06.cap)

WPA PSK handshake capture

**Comment by aircrack-ng** [comment]
**Attachment by hvidevold on 26 Aug 2008 02:35 UTC**

Uploaded file [testwords.txt](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20080826-023534_000-testwords.txt)

Wordlist with correct password for airodump-06.cap

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2007 14:25 UTC**

See also #412, #522, #515

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

I have the fix for this test case. I ran into this same bug when playing with cowpatty 4.3 and aircrack 1.0-rc1. The fix is simple, stop trying to process two keys per loop. Instead just do one key per loop. I will make an attachment of my patch.

**Comment by aircrack-ng** [comment]
**Attachment by edgan on 30 Oct 2008 04:10 UTC**

Uploaded file [aircrack-ng-1.0rc2-one-at-a-time.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081030-041006_000-aircrack-ng-1.0rc2-one-at-a-time.patch)

Patch to fix only one string in the wordlist problem

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

I have fixed a bug in my patch that caused it to report the keys per second at double what they actually were. I also in the process fixed a few typos on the word dictionary(was dictionnary).

**Comment by aircrack-ng** [comment]
**Attachment by edgan on 30 Oct 2008 19:44 UTC**

Uploaded file [aircrack-ng-1.0rc2-one-at-a-time-v2.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081030-194416_000-aircrack-ng-1.0rc2-one-at-a-time-v2.patch)

Improved version of the one at a time patch that includes a k/s bugfix

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

I have improved my one at a time patch even further. 

Before two keys per thread/cpu were done at a time, which did improve performance. It also lead to this bug report. My original patches reduced it to one key per thread/cpu, but that seemed to cause performance to be intermittent.

So I modified the patch to do, if WPA, then opt.nbcpu = opt.nbcpu * 2, which then uses twice as many threads. This brings performance back to the level it was before, but doesn't reintroduce the same bug.

I will attach my latest patch.

**Comment by aircrack-ng** [comment]
**Attachment by edgan on 31 Oct 2008 07:21 UTC**

Uploaded file [aircrack-ng-1.0rc2-one-at-a-time-v3.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081031-072111_000-aircrack-ng-1.0rc2-one-at-a-time-v3.patch)

Improved version of the one at a time patch that brings performance back to the pre-patch level

**Comment by aircrack-ng** [comment]
**Modified by misterx on 16 Feb 2007 14:25 UTC**

**Comment by aircrack-ng** [comment]
**Attachment by misterx on 11 Nov 2008 19:42 UTC**

Uploaded file [aircrack-ng-1.0rc2-one-at-a-time-v4.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081111-194237_000-aircrack-ng-1.0rc2-one-at-a-time-v4.patch)

Updated v3 (spelling already corrected in svn).

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

v4 seems to work for me. It works with a one word dictionary, and the key rate seems to be what it should be.

**Comment by aircrack-ng** [comment]
**Attachment by misterx on 13 Nov 2008 23:17 UTC**

Uploaded file [aircrack-ng-1.0rc2-one-at-a-time-v5.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081113-231714_000-aircrack-ng-1.0rc2-one-at-a-time-v5.patch)

Fixed the number of passphrases tried.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 16 Feb 2007 14:25 UTC**

Tried the aircrack-ng-1.0rc2-one-at-a-time-v5.patch with svn revision 1225.

aircrack-ng will stop testing after 1126 passphrases.

Tested this together with the Church of Wifi 40Gb WPA-tables. And if I do:

airolib-ng <database> --stats

I see the SSID is linked to ~999000 passphrases.

So still not fixed!

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 16 Feb 2007 14:25 UTC**

Also -p 1 does not work for me. Also stops at 1126 passphrases.

Only goes slower because of only 1 core used.

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

Please provide exact common line options used, and the handshake file if you can. Also how are you using the WPA tables with aircrack-ng? They are meant for cowpatty. Did you convert them into airolib-ng format, aka sqlite?

**Comment by aircrack-ng** [comment]
**Attachment by Nick on 21 Nov 2008 15:57 UTC**

Uploaded file [passphrases tried.jpg](aircrack-ng)(https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081121-155734_000-aircrack-ng%20passphrases%20tried.jpg)

number of passphrases tried

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 16 Feb 2007 14:25 UTC**

Running under Backtrack 3 Final on my laptop
--------------------------------------------

First I downloaded aircrack-ng svn 1225:

'''''svn co http://trac.aircrack-ng.org/svn/trunk aircrack-ng'''''

Second I applied the aircrack-ng-1.0rc2-one-at-a-time-v5.patch to the downloaded aircrack:

'''''patch -i aircrack-ng-1.0rc2-one-at-a-time-v5.patch <"path to aircrack-ng.c">'''''

Third I installed the patched aircrack-ng 1225:

'''''cd aircrack-ng
make
make install'''''

Fourth I convert the WPA-table according to SSID name to the airolib-ng format:

'''''airolib-ng testdb --import cowpatty <"path to ssid name file">'''''

After the above command I type:

'''''airolib-ng testdb --stats'''''

And it outputs the following:

'''''There are 1 ESSIDs and 995759 passwords in the database. 995759 out of 995759 possible combinations have been computed (100%).'''''

Fifth I start airodump-ng to focus on my AP and to capture the handshake:

'''''airodump-ng -c 4 -w test -b <mac address of my AP> wlan1'''''

Sixth I use aireplay-ng to deauthenticate my desktop pc witch is associated with my AP:

'''''aireplay-ng -0 1 -a <mac address of my AP> -c <mac address of my desktop pc> wlan1'''''

After the above command I can clearly see that airodump-ng has captured the WPA-handshake in the top-right corner (see attachment "test-01.cap")

Seventh I run aircrack-ng to crack the passphrase using the created databse by airolib-ng and the created capture file by airodump-ng:

'''''aircrack-ng -w testdb -b <mac address of my AP> test-01.cap'''''

After entering the above command aircrack-ng stops at exactly 1125 or 1126 passphrases (one time it's 1125 the other time it's 1126) Even if I use the -p option it stops at these numbers and it outputs:

'''''Passphrase is not in dictionary.'''''

Altough I use a password I copied from the password list the CoW WPA tables use.
For the full output of aircrack-ng see "aircrack-ng passphrases tried.jpg".
As you can see I also have to press Ctrl + C a couple of times before aircrack-ng actualy quits correctly, but this issue is already discribed in another ticket.

Kind regards.
Nick

**Comment by aircrack-ng** [comment]
**Attachment by Nick on 21 Nov 2008 18:52 UTC**

Uploaded file [test-01.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081121-185211_000-test-01.cap)

**Comment by aircrack-ng** [comment]
**Comment by edgan on 16 Feb 2007 14:25 UTC**

I do get the same type of results when I do the same thing, but you are doing it wrong. You are using -w testdb, when you should be using -r testdb.

airolib-ng testdb2 --stats                                            
There are 1 ESSIDs and 22353 passwords in the database. 22353 out of 22353 possible combinations have been computed (100%).


aircrack-ng -w testdb2 /usr/share/doc/cowpatty-4.3/wpapsk-linksys.dump

[30 keys tested (365.23 k/s)

Passphrase not in dictionary 


aircrack-ng -r testdb2 /usr/share/doc/cowpatty-4.3/wpapsk-linksys.dump

[00:00:00](00:00:00]) 22352 keys tested (70756.57 k/s)

KEY FOUND! [ dictionary ]

**Comment by aircrack-ng** [comment]
**Attachment by edgan on 22 Nov 2008 20:50 UTC**

Uploaded file [aircrack-ng-1.0rc2-off-by-one.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081122-205011_000-aircrack-ng-1.0rc2-off-by-one.patch)

Patch for one-at-a-time-v5 to fix an off by one bug on the tested key count

**Comment by aircrack-ng** [comment]
**Attachment by edgan on 23 Nov 2008 02:48 UTC**

Uploaded file [aircrack-ng-1.0rc2-wordlist.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081123-024844_000-aircrack-ng-1.0rc2-wordlist.patch)

Patch to add detection to -w of files that aren't proper word lists

**Comment by aircrack-ng** [comment]
**Attachment by misterx on 29 Nov 2008 02:22 UTC**

Uploaded file [aircrack-ng-1.0rc2-wordlist_v2.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/109/20081129-022211_000-aircrack-ng-1.0rc2-wordlist_v2.patch)

Updated patch (and a few fixes)

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2007 14:25 UTC**

Fixed in aircrack-ng/aircrack-ng@8381e592f3e4372b457dc5ff6a1d12122309da71
